File: src/python/jterator/api/io.py
from __future__ import unicode_literals
import sys
import re
import yaml
import h5py
from jterator.error import JteratorError

# from IPython.core.debugger import Tracer


def gethandles(handles_stream):
    '''
    Reading "handles" from YAML file.
    '''
    mfilename = sys._getframe().f_code.co_name
    handles = yaml.load(handles_stream)

    print('jt -- %s: loaded \'handles\'' % mfilename)

    return handles


def readinputargs(handles):
    '''
    Reading input arguments from HDF5 file
    using the location specified in "handles".
    '''
    mfilename = sys._getframe().f_code.co_name

    hdf5_filename = handles['hdf5_filename']
    hdf5_root = h5py.File(hdf5_filename, 'r+')

    input_args = dict()
    for key in handles['input_keys']:
        field = handles['input_keys'][key]
        input_args[key] = dict()

        if 'hdf5_location' in field:
            # note: we index into the dataset to retrieve its content,
            # otherwise it would be loaded as hdf5 object
            input_args[key]['variable'] = hdf5_root[field['hdf5_location']][()]
            print('jt -- %s: loaded dataset \'%s\' from HDF5 group: "%s"'
                  % (mfilename, key, field['hdf5_location']))
        elif 'parameter' in field:
            input_args[key]['variable'] = field['parameter']
            print('jt -- %s: parameter \'%s\': "%s"'
                  % (mfilename, key, str(field['parameter'])))
        else:
            raise JteratorError('Possible variable keys are '
                                '\'hdf5_location\' or \'parameter\'')

        if 'class' in field:
            input_args[key]['class'] = field['class']

        if 'attributes' in field:
            input_args[key]['attributes'] = field['attributes']

    h5py.File.close(hdf5_root)

    return input_args


def checkinputargs(input_args):
    '''
    Checks input arguments for correct class (i.e. type) and attributes
    (attributes are not yet implemented).
    '''
    mfilename = sys._getframe().f_code.co_name

    checked_input_args = dict()
    for key in input_args:

        # checks are only done if "class" is specified
        if 'class' in input_args[key]:
            expected_class = input_args[key]['class']
            loaded_class = type(input_args[key]['variable'])

            if 'h5py' in str(loaded_class):
                loaded_class = input_args[key]['variable'].dtype
            else:
                loaded_class = loaded_class.__name__

            if str(loaded_class) != expected_class:
                raise JteratorError('argument \'%s\' is of "class" \'%s\' '
                                    'instead of expected \'%s\''
                                    % (key, loaded_class, expected_class))

            print('jt -- %s: argument \'%s\' passed check' % (mfilename, key))

        else:
            print('jt -- %s: argument \'%s\' not checked' % (mfilename, key))

        # return parameters in simplified form
        checked_input_args[key] = input_args[key]['variable']

    return checked_input_args


def writeoutputargs(handles, output_args):
    '''
    Writing output arguments to HDF5 file
    using the location specified in "handles".
    '''
    mfilename = sys._getframe().f_code.co_name

    hdf5_filename = hdf5_filename = re.sub('/tmp/(.*)\.tmp$', '/data/\\1.data',
                           handles['hdf5_filename'])
    hdf5_root = h5py.File(hdf5_filename, 'r+')

    for key in output_args:
        hdf5_location = handles['output_keys'][key]['hdf5_location']
        hdf5_root.create_dataset(hdf5_location, data=output_args[key])
        print('jt -- %s: wrote dataset \'%s\' to HDF5 group: "%s"'
              % (mfilename, key, hdf5_location))

    h5py.File.close(hdf5_root)


def writeoutputtmp(handles, output_tmp):
    '''
    Writing output arguments to HDF5 file
    using the location specified in "handles".
    '''
    mfilename = sys._getframe().f_code.co_name

    hdf5_filename = handles['hdf5_filename']
    hdf5_root = h5py.File(hdf5_filename, 'r+')

    for key in output_tmp:
        hdf5_location = handles['output_keys'][key]['hdf5_location']
        hdf5_root.create_dataset(hdf5_location, data=output_tmp[key])
        print('jt -- %s: wrote tmp dataset \'%s\' to HDF5 group: "%s"'
              % (mfilename, key, hdf5_location))

    h5py.File.close(hdf5_root)


def buildhdf5(handles):
    '''
    Create HDF5 file.
    '''
    mfilename = sys._getframe().f_code.co_name

    hdf5_filename = handles['hdf5_filename']
    h5py.File(hdf5_filename, 'w')
    print('jt -- %s: created HDF5 file for temporary data: "%s"'
          % (mfilename, hdf5_filename))

    hdf5_filename = re.sub('/tmp/(.*)\.tmp$', '/data/\\1.data',
                           handles['hdf5_filename'])
    h5py.File(hdf5_filename, 'w')
    print('jt -- %s: created HDF5 file for measurement pipe data: "%s"'
          % (mfilename, hdf5_filename))

File: src/julia/jterator/api/io.jl
module jterator

import YAML
import HDF5, JLD

###################################################################
## Note: the HDF5 package doesn't handle dimensions correctly!!! ##
###################################################################

function gethandles(handles_stream)
    ## Reading "handles" from YAML file.

    mfilename = "gethandles" # can we get this automatically?

    # Reading handles from YAML.
    handles = YAML.load(handles_stream)

    @printf("jt -- %s: loaded handles\n", mfilename)

    return handles

end


function readinputargs(handles)
    ## Reading input arguments from HDF5 file
    ## using the location specified in "handles".

    mfilename = "readinputargs"
    
    hdf5_filename = handles["hdf5_filename"]

    input_args = Dict()
    for key in keys(handles["input"])
        field = handles["input"][key]
        input_args[key] = Dict()

        if haskey(field, "hdf5_location")
            input_args[key]["variable"] = HDF5.h5read(hdf5_filename, field["hdf5_location"])
            if ndims(input_args[key]["variable"]) > 1
                input_args[key]["variable"] = input_args[key]["variable"]'
            end
            @printf("jt -- %s: loaded dataset '%s' from HDF5 group: \"%s\"\n", mfilename, key, field["hdf5_location"])
        elseif haskey(field, "parameter")
            input_args[key]["variable"] = field["parameter"]
            @printf("jt -- %s: parameter '%s': \"%s\"\n", mfilename, key, input_args[key]["variable"])          
        else
            error("Possible variable keys are \"hdf5_location\" or \"parameter\"")
        end 

        if haskey(field, "class")
            input_args[key]["class"] = field["class"]
        end 

    end

    return input_args
    
end


function checkinputargs(input_args)
    ##Checks input arguments for correct class and attributes.
    
    mfilename = "checkinputargs"

    checked_input_args = Dict()
    for key in keys(input_args)

        # checks are only done if "class" is specified
        if haskey(input_args[key], "class")
            expected_class = input_args[key]["class"]
            loaded_class = typeof(input_args[key]["variable"])

            if ~isequal(string(loaded_class), expected_class) 
                error(@sprintf("argument \"%s\" is of class \"%s\" instead of expected \"%s\"", 
                      key, loaded_class, expected_class))
            end

            @printf("jt -- %s: argument \"%s\" passed check\n", mfilename, key)

        else
            @printf("jt -- %s: argument \"%s\" not checked\n", mfilename, key)
        end

        # return parameters in simplified form
        checked_input_args[key] = input_args[key]["variable"]

    end

    return checked_input_args

end


function writedata(handles, data)
    ##Writing data to HDF5 file.

    mfilename = "writedata"

    hdf5_filename = handles["hdf5_filename"]
    orig_substr = match(r"tmp/(.*)\.tmp", hdf5_filename).captures[1]
    hdf5_filename = replace(hdf5_filename, r"/tmp/(.*)\.tmp$", 
                            @sprintf("/data/%s.data", orig_substr))
    hdf5_root = HDF5.h5open(hdf5_filename, "r+")

    for key in keys(data)
        hdf5_location = handles["output"][key]["hdf5_location"]
        if ndims(output_args[key]) > 1
            hdf5_root[hdf5_location] = data[key]'
        else
            hdf5_root[hdf5_location] = data[key]
        end
        @printf("jt -- %s: wrote dataset '%s' to HDF5 group: \"%s\"\n",
                mfilename, key, hdf5_location)
    end

    close(hdf5_root)

end


function writeoutputargs(handles, output_args)
    ## Writing output arguments to HDF5 file
    ## using the location specified in "handles".

    mfilename = "writeoutputargs"

    hdf5_filename = handles["hdf5_filename"]
    hdf5_root = HDF5.h5open(hdf5_filename, "r+")

    for key in keys(output_args)
        hdf5_location = handles["output"][key]["hdf5_location"]
        if ndims(output_args[key]) > 1
            hdf5_root[hdf5_location] = output_args[key]'
        else
            hdf5_root[hdf5_location] = output_args[key]
        end
        @printf("jt -- %s: wrote tmp dataset '%s' to HDF5 group: \"%s\"\n",
                mfilename, key, hdf5_location)
    end

    close(hdf5_root)

end


export gethandles, readinputargs, checkinputargs, writeoutputargs, writedata

end
